File: DSA-C++/6-array/9-union-of-sorted-arrays.cpp
#include <bits/stdc++.h>

using namespace std;
// Problem Statement: Given two sorted arrays, arr1, and arr2 of size n and m. Find the union of two sorted arrays.
// The union of two arrays can be defined as the common and distinct elements in the two arrays.
// NOTE: Elements in the union should be in ascending order.



// Ordered set - O(n.logn) , O(n) 
vector <int> union_of_sorted_arrays_1(vector <int> &nums1, vector <int> &nums2) {
    set <int> set(nums1.begin(), nums1.end());
    vector <int> temp;

    for (auto x:nums2) {
        set.insert(x);
    }

    int i=0;
    for (auto x:set) {
        temp.push_back(x);
    }

    return temp;
}

// Two pointers approach - O(n) , O(n)
vector <int> union_of_sorted_arrays_2(vector <int> &arr1, vector <int> &arr2) {
    int l1=0 , l2=0;
    vector <int> temp;

    while (l1 < arr1.size() && l2 < arr2.size()) {
        if (arr1[l1] < arr2[l2]) {
            if (temp.empty() || temp.back() != arr1[l1]) {
                temp.push_back(arr1[l1]);
            }
            l1++;
        }

        else if (arr1[l1] > arr2[l2]) {
            if (temp.empty() || temp.back() != arr2[l2]) {
                temp.push_back(arr2[l2]);
            }

            l2++;
        }

        else {
            if (temp.empty() || temp.back() != arr1[l1]) {
                temp.push_back(arr1[l1]);
            }
            l1++;
            l2++;
        }
    }

    while (l1 < arr1.size()) {
        if (temp.empty() || temp.back() != arr1[l1]) {
            temp.push_back(arr1[l1]);
        }
        l1++;
    }

    while (l2 < arr2.size()) {
        if (temp.empty() || temp.back() != arr2[l2]) {
            temp.push_back(arr2[l2]);
        }

        l2++;
    }

    return temp;
}


int main() {
    vector <int> arr1 = {1,2,3,4,5,6,7};
    vector <int> arr2 = {-1,-100,3,99};

    vector <int> ans = union_of_sorted_arrays_1(arr1,arr2);

    // vector <int> ans = union_of_sorted_arrays_2(arr1,arr2);

    for (auto x:ans) {
        cout << x << " ";
    }

    cout << endl;
    return 0;
}

File: DSA-Python/5-array/9-union-of-sorted-arrays.py
# Given two sorted arrays, arr1, and arr2 of size n and m. Find the union of two sorted arrays.
# The union of two arrays can be defined as the common and distinct elements in the two arrays.
# NOTE: Elements in the union should be in ascending order.


# hashing - O(k.log(k)), O(n) 
# k is total unique elements accross both arrays
def union_of_sorted_arrays_1(nums1:list[int], nums2:list[int]) -> list[int]:
    uset1 = set(nums1)
    uset2 = set(nums2)
    
    return sorted(uset1.union(uset2))
    
    
    
    
# two pointers - O(n), O(1)
def union_of_sorted_arrays_2(nums1:list[int], nums2:list[int]) -> list[int]:
    n1, n2 = len(nums1), len(nums2)
    l1, l2 = 0, 0
    ans = []
    
    while l1 < n1 and l2 < n2:
        if nums1[l1] < nums2[l2]:
            if not ans or ans[-1] != nums1[l1]:
                ans.append(nums1[l1])
            l1 += 1
        elif nums1[l1] > nums2[l2]:
            if not ans or nums2[l2] != ans[-1]:
                ans.append(nums2[l2])
            l2 += 1
        else:
            if not ans or nums1[l1] != ans[-1]:
                ans.append(nums2[l2])
            l1 += 1
            l2 += 1 
            
    
    while l1 < n1:
        if not ans or nums1[l1] != ans[-1]:
            ans.append(nums1[l1])
        l1 += 1
            
        
    while l2 < n2:
        if not ans or nums2[l2] != ans[-1]:
            ans.append(nums2[l2])
        l2 += 1

    return ans

    
    
if __name__ == "__main__": 
    loops = int(input())

    for _ in range(loops):
        nums1 = list(map(int, input().split()))
        nums2 = list(map(int, input().split()))
        
        # ans = union_of_sorted_arrays_1(nums1, nums2)
        ans = union_of_sorted_arrays_2(nums1, nums2)
        
        print(*ans)
        
'''
4
0 1 2 3 4 5 6 8
-100 -1 3 99
0
-4 -3 -2 -1

-1
1 2 3 4 5 6
3 4 5 6 7 8
'''